File: sdk/javascript/src/template.js
/**
 * FoodBlock Templates — reusable patterns for common workflows.
 * Templates are themselves FoodBlocks (observe.template) that describe
 * a sequence of block types with expected state and ref relationships.
 */

const { create } = require('./block')

/**
 * Create a template block.
 * @param {string} name - Template name
 * @param {string} description - What the template models
 * @param {Array} steps - Array of step definitions
 * @param {object} [opts] - { author: hash }
 * @returns {object} The template FoodBlock
 */
function createTemplate(name, description, steps, opts = {}) {
  const refs = {}
  if (opts.author) refs.author = opts.author
  return create('observe.template', {
    name,
    description,
    steps
  }, refs)
}

/**
 * Instantiate a template — create real blocks from a template pattern.
 *
 * @param {object} template - A template block (or its state)
 * @param {object} values - Map of step alias to { state, refs } overrides
 * @param {object} [opts] - { registry } for alias resolution
 * @returns {object[]} Array of created blocks, in dependency order
 */
function fromTemplate(template, values = {}) {
  const state = template.state || template
  const steps = state.steps
  if (!steps || !Array.isArray(steps)) {
    throw new Error('FoodBlock: template must have steps array')
  }

  const aliases = new Map()
  const blocks = []

  for (const step of steps) {
    const alias = step.alias || step.type
    const overrides = values[alias] || {}

    // Build state from step defaults + overrides
    const blockState = { ...(step.default_state || {}) }
    if (Array.isArray(step.required)) {
      for (const field of step.required) {
        if (!overrides.state || !(field in overrides.state)) {
          // Skip if required field not provided (partial instantiation)
        }
      }
    }
    if (overrides.state) Object.assign(blockState, overrides.state)

    // Build refs, resolving @aliases to previously created block hashes
    const blockRefs = {}
    if (step.refs) {
      for (const [role, target] of Object.entries(step.refs)) {
        if (typeof target === 'string' && target.startsWith('@')) {
          const refAlias = target.slice(1)
          if (aliases.has(refAlias)) {
            blockRefs[role] = aliases.get(refAlias)
          }
        } else {
          blockRefs[role] = target
        }
      }
    }
    // Override refs from values
    if (overrides.refs) {
      for (const [role, target] of Object.entries(overrides.refs)) {
        if (typeof target === 'string' && target.startsWith('@')) {
          const refAlias = target.slice(1)
          if (aliases.has(refAlias)) {
            blockRefs[role] = aliases.get(refAlias)
          }
        } else {
          blockRefs[role] = target
        }
      }
    }

    const block = create(step.type, blockState, blockRefs)
    aliases.set(alias, block.hash)
    blocks.push(block)
  }

  return blocks
}

/**
 * Built-in templates for common patterns.
 */
const TEMPLATES = {
  'supply-chain': {
    name: 'Farm-to-Table Supply Chain',
    description: 'A complete provenance chain from primary producer to retail',
    steps: [
      { type: 'actor.producer', alias: 'farm', required: ['name'] },
      { type: 'substance.ingredient', alias: 'crop', refs: { source: '@farm' }, required: ['name'] },
      { type: 'transform.process', alias: 'processing', refs: { input: '@crop' }, required: ['name'] },
      { type: 'substance.product', alias: 'product', refs: { origin: '@processing' }, required: ['name'] },
      { type: 'transfer.order', alias: 'sale', refs: { item: '@product' } }
    ]
  },
  'review': {
    name: 'Product Review',
    description: 'A consumer review of a food product',
    steps: [
      { type: 'actor.venue', alias: 'venue', required: ['name'] },
      { type: 'substance.product', alias: 'product', refs: { seller: '@venue' }, required: ['name'] },
      { type: 'observe.review', alias: 'review', refs: { subject: '@product' }, required: ['rating'] }
    ]
  },
  'certification': {
    name: 'Product Certification',
    description: 'An authority certifying a producer or product',
    steps: [
      { type: 'actor.authority', alias: 'authority', required: ['name'] },
      { type: 'actor.producer', alias: 'producer', required: ['name'] },
      { type: 'observe.certification', alias: 'cert', refs: { authority: '@authority', subject: '@producer' }, required: ['name'] }
    ]
  },
  'surplus-rescue': {
    name: 'Surplus Rescue',
    description: 'Food business posts surplus, sustainer collects, donation recorded',
    steps: [
      { type: 'actor.venue', alias: 'donor', required: true, defaultState: { name: 'Food Business' } },
      { type: 'substance.surplus', alias: 'surplus', refs: { seller: '@donor' }, required: true, defaultState: { name: 'Surplus Food', status: 'available' } },
      { type: 'transfer.donation', alias: 'donation', refs: { source: '@donor', item: '@surplus' }, required: true, defaultState: { status: 'collected' } }
    ]
  },
  'agent-reorder': {
    name: 'Agent Reorder',
    description: 'Inventory check → low stock → draft order → approve → order placed',
    steps: [
      { type: 'actor.venue', alias: 'business', required: true, defaultState: { name: 'Business' } },
      { type: 'observe.reading', alias: 'inventory-check', refs: { subject: '@business' }, required: true, defaultState: { name: 'Inventory Check', reading_type: 'stock_level' } },
      { type: 'actor.agent', alias: 'agent', refs: { operator: '@business' }, required: true, defaultState: { name: 'Reorder Agent', capabilities: ['ordering'] } },
      { type: 'transfer.order', alias: 'draft-order', refs: { buyer: '@business', agent: '@agent' }, required: true, defaultState: { status: 'draft', draft: true } },
      { type: 'transfer.order', alias: 'confirmed-order', refs: { buyer: '@business', updates: '@draft-order' }, required: true, defaultState: { status: 'confirmed' } }
    ]
  },
  'restaurant-sourcing': {
    name: 'Restaurant Sourcing',
    description: 'Restaurant needs ingredient → discovery → supplier offer → accept → order → delivery',
    steps: [
      { type: 'actor.venue', alias: 'restaurant', required: true, defaultState: { name: 'Restaurant' } },
      { type: 'substance.ingredient', alias: 'needed', refs: {}, required: true, defaultState: { name: 'Ingredient Needed' } },
      { type: 'actor.producer', alias: 'supplier', required: true, defaultState: { name: 'Supplier' } },
      { type: 'transfer.offer', alias: 'offer', refs: { seller: '@supplier', item: '@needed', buyer: '@restaurant' }, required: true, defaultState: { status: 'offered' } },
      { type: 'transfer.order', alias: 'order', refs: { buyer: '@restaurant', seller: '@supplier', item: '@needed' }, required: true, defaultState: { status: 'confirmed' } },
      { type: 'transfer.delivery', alias: 'delivery', refs: { order: '@order', seller: '@supplier', buyer: '@restaurant' }, required: true, defaultState: { status: 'delivered' } }
    ]
  },
  'food-safety-audit': {
    name: 'Food Safety Audit',
    description: 'Inspector visits → readings taken → report → certification → attestation',
    steps: [
      { type: 'actor.venue', alias: 'premises', required: true, defaultState: { name: 'Food Premises' } },
      { type: 'actor.producer', alias: 'inspector', required: true, defaultState: { name: 'Food Safety Inspector' } },
      { type: 'observe.reading', alias: 'readings', refs: { subject: '@premises', author: '@inspector' }, required: true, defaultState: { name: 'Safety Readings' } },
      { type: 'observe.certification', alias: 'certificate', refs: { subject: '@premises', authority: '@inspector' }, required: true, defaultState: { name: 'Food Safety Certificate' } },
      { type: 'observe.attestation', alias: 'attestation', refs: { confirms: '@certificate', attestor: '@inspector' }, required: true, defaultState: { confidence: 'verified' } }
    ]
  },
  'market-day': {
    name: 'Market Day',
    description: 'Producer brings stock → stall setup → sales → end-of-day surplus → donation',
    steps: [
      { type: 'actor.producer', alias: 'producer', required: true, defaultState: { name: 'Market Producer' } },
      { type: 'place.market', alias: 'market', required: true, defaultState: { name: 'Farmers Market' } },
      { type: 'substance.product', alias: 'stock', refs: { seller: '@producer' }, required: true, defaultState: { name: 'Market Stock' } },
      { type: 'transfer.order', alias: 'sales', refs: { seller: '@producer', item: '@stock' }, required: false, defaultState: { status: 'completed' } },
      { type: 'substance.surplus', alias: 'leftover', refs: { seller: '@producer', source: '@stock' }, required: false, defaultState: { name: 'End of Day Surplus', status: 'available' } }
    ]
  },
  'cold-chain': {
    name: 'Cold Chain',
    description: 'Shipment departs → temperature readings → delivery → chain verified',
    steps: [
      { type: 'actor.distributor', alias: 'carrier', required: true, defaultState: { name: 'Cold Chain Carrier' } },
      { type: 'transfer.delivery', alias: 'shipment', refs: { carrier: '@carrier' }, required: true, defaultState: { status: 'in_transit' } },
      { type: 'observe.reading', alias: 'temp-log', refs: { subject: '@shipment' }, required: true, defaultState: { name: 'Temperature Log', reading_type: 'temperature' } },
      { type: 'observe.attestation', alias: 'chain-verified', refs: { confirms: '@shipment', attestor: '@carrier' }, required: true, defaultState: { confidence: 'verified', method: 'continuous_monitoring' } }
    ]
  }
}

module.exports = { createTemplate, fromTemplate, TEMPLATES }

File: sdk/go/template.go
package foodblock

// TemplateStep defines a single step in a template.
type TemplateStep struct {
	Type         string            `json:"type"`
	Alias        string            `json:"alias,omitempty"`
	Refs         map[string]string `json:"refs,omitempty"`
	Required     []string          `json:"required,omitempty"`
	DefaultState map[string]interface{} `json:"default_state,omitempty"`
}

// TemplateDef defines a reusable block creation pattern.
type TemplateDef struct {
	Name        string         `json:"name"`
	Description string         `json:"description"`
	Steps       []TemplateStep `json:"steps"`
}

// StepOverrides provides state and ref overrides for template instantiation.
type StepOverrides struct {
	State map[string]interface{}
	Refs  map[string]string
}

// Templates is the set of built-in template definitions.
var Templates = map[string]TemplateDef{
	"supply-chain": {
		Name:        "Farm-to-Table Supply Chain",
		Description: "A complete provenance chain from primary producer to retail",
		Steps: []TemplateStep{
			{Type: "actor.producer", Alias: "farm", Required: []string{"name"}},
			{Type: "substance.ingredient", Alias: "crop", Refs: map[string]string{"source": "@farm"}, Required: []string{"name"}},
			{Type: "transform.process", Alias: "processing", Refs: map[string]string{"input": "@crop"}, Required: []string{"name"}},
			{Type: "substance.product", Alias: "product", Refs: map[string]string{"origin": "@processing"}, Required: []string{"name"}},
			{Type: "transfer.order", Alias: "sale", Refs: map[string]string{"item": "@product"}},
		},
	},
	"review": {
		Name:        "Product Review",
		Description: "A consumer review of a food product",
		Steps: []TemplateStep{
			{Type: "actor.venue", Alias: "venue", Required: []string{"name"}},
			{Type: "substance.product", Alias: "product", Refs: map[string]string{"seller": "@venue"}, Required: []string{"name"}},
			{Type: "observe.review", Alias: "review", Refs: map[string]string{"subject": "@product"}, Required: []string{"rating"}},
		},
	},
	"certification": {
		Name:        "Product Certification",
		Description: "An authority certifying a producer or product",
		Steps: []TemplateStep{
			{Type: "actor.authority", Alias: "authority", Required: []string{"name"}},
			{Type: "actor.producer", Alias: "producer", Required: []string{"name"}},
			{Type: "observe.certification", Alias: "cert", Refs: map[string]string{"authority": "@authority", "subject": "@producer"}, Required: []string{"name"}},
		},
	},
	"surplus-rescue": {
		Name:        "Surplus Rescue",
		Description: "Food business posts surplus, sustainer collects, donation recorded",
		Steps: []TemplateStep{
			{Type: "actor.venue", Alias: "donor", DefaultState: map[string]interface{}{"name": "Food Business"}},
			{Type: "substance.surplus", Alias: "surplus", Refs: map[string]string{"seller": "@donor"}, DefaultState: map[string]interface{}{"name": "Surplus Food", "status": "available"}},
			{Type: "transfer.donation", Alias: "donation", Refs: map[string]string{"source": "@donor", "item": "@surplus"}, DefaultState: map[string]interface{}{"status": "collected"}},
		},
	},
	"agent-reorder": {
		Name:        "Agent Reorder",
		Description: "Inventory check → low stock → draft order → approve → order placed",
		Steps: []TemplateStep{
			{Type: "actor.venue", Alias: "business", DefaultState: map[string]interface{}{"name": "Business"}},
			{Type: "observe.reading", Alias: "inventory-check", Refs: map[string]string{"subject": "@business"}, DefaultState: map[string]interface{}{"name": "Inventory Check", "reading_type": "stock_level"}},
			{Type: "actor.agent", Alias: "agent", Refs: map[string]string{"operator": "@business"}, DefaultState: map[string]interface{}{"name": "Reorder Agent", "capabilities": []interface{}{"ordering"}}},
			{Type: "transfer.order", Alias: "draft-order", Refs: map[string]string{"buyer": "@business", "agent": "@agent"}, DefaultState: map[string]interface{}{"status": "draft", "draft": true}},
			{Type: "transfer.order", Alias: "confirmed-order", Refs: map[string]string{"buyer": "@business", "updates": "@draft-order"}, DefaultState: map[string]interface{}{"status": "confirmed"}},
		},
	},
	"restaurant-sourcing": {
		Name:        "Restaurant Sourcing",
		Description: "Restaurant needs ingredient → discovery → supplier offer → accept → order → delivery",
		Steps: []TemplateStep{
			{Type: "actor.venue", Alias: "restaurant", DefaultState: map[string]interface{}{"name": "Restaurant"}},
			{Type: "substance.ingredient", Alias: "needed", Refs: map[string]string{}, DefaultState: map[string]interface{}{"name": "Ingredient Needed"}},
			{Type: "actor.producer", Alias: "supplier", DefaultState: map[string]interface{}{"name": "Supplier"}},
			{Type: "transfer.offer", Alias: "offer", Refs: map[string]string{"seller": "@supplier", "item": "@needed", "buyer": "@restaurant"}, DefaultState: map[string]interface{}{"status": "offered"}},
			{Type: "transfer.order", Alias: "order", Refs: map[string]string{"buyer": "@restaurant", "seller": "@supplier", "item": "@needed"}, DefaultState: map[string]interface{}{"status": "confirmed"}},
			{Type: "transfer.delivery", Alias: "delivery", Refs: map[string]string{"order": "@order", "seller": "@supplier", "buyer": "@restaurant"}, DefaultState: map[string]interface{}{"status": "delivered"}},
		},
	},
	"food-safety-audit": {
		Name:        "Food Safety Audit",
		Description: "Inspector visits → readings taken → report → certification → attestation",
		Steps: []TemplateStep{
			{Type: "actor.venue", Alias: "premises", DefaultState: map[string]interface{}{"name": "Food Premises"}},
			{Type: "actor.producer", Alias: "inspector", DefaultState: map[string]interface{}{"name": "Food Safety Inspector"}},
			{Type: "observe.reading", Alias: "readings", Refs: map[string]string{"subject": "@premises", "author": "@inspector"}, DefaultState: map[string]interface{}{"name": "Safety Readings"}},
			{Type: "observe.certification", Alias: "certificate", Refs: map[string]string{"subject": "@premises", "authority": "@inspector"}, DefaultState: map[string]interface{}{"name": "Food Safety Certificate"}},
			{Type: "observe.attestation", Alias: "attestation", Refs: map[string]string{"confirms": "@certificate", "attestor": "@inspector"}, DefaultState: map[string]interface{}{"confidence": "verified"}},
		},
	},
	"market-day": {
		Name:        "Market Day",
		Description: "Producer brings stock → stall setup → sales → end-of-day surplus → donation",
		Steps: []TemplateStep{
			{Type: "actor.producer", Alias: "producer", DefaultState: map[string]interface{}{"name": "Market Producer"}},
			{Type: "place.market", Alias: "market", DefaultState: map[string]interface{}{"name": "Farmers Market"}},
			{Type: "substance.product", Alias: "stock", Refs: map[string]string{"seller": "@producer"}, DefaultState: map[string]interface{}{"name": "Market Stock"}},
			{Type: "transfer.order", Alias: "sales", Refs: map[string]string{"seller": "@producer", "item": "@stock"}, DefaultState: map[string]interface{}{"status": "completed"}},
			{Type: "substance.surplus", Alias: "leftover", Refs: map[string]string{"seller": "@producer", "source": "@stock"}, DefaultState: map[string]interface{}{"name": "End of Day Surplus", "status": "available"}},
		},
	},
	"cold-chain": {
		Name:        "Cold Chain",
		Description: "Shipment departs → temperature readings → delivery → chain verified",
		Steps: []TemplateStep{
			{Type: "actor.distributor", Alias: "carrier", DefaultState: map[string]interface{}{"name": "Cold Chain Carrier"}},
			{Type: "transfer.delivery", Alias: "shipment", Refs: map[string]string{"carrier": "@carrier"}, DefaultState: map[string]interface{}{"status": "in_transit"}},
			{Type: "observe.reading", Alias: "temp-log", Refs: map[string]string{"subject": "@shipment"}, DefaultState: map[string]interface{}{"name": "Temperature Log", "reading_type": "temperature"}},
			{Type: "observe.attestation", Alias: "chain-verified", Refs: map[string]string{"confirms": "@shipment", "attestor": "@carrier"}, DefaultState: map[string]interface{}{"confidence": "verified", "method": "continuous_monitoring"}},
		},
	},
}

// CreateTemplate creates an observe.template FoodBlock.
func CreateTemplate(name, description string, steps []TemplateStep, authorHash string) Block {
	stepsSlice := make([]interface{}, len(steps))
	for i, s := range steps {
		step := map[string]interface{}{"type": s.Type}
		if s.Alias != "" {
			step["alias"] = s.Alias
		}
		if len(s.Refs) > 0 {
			refs := make(map[string]interface{})
			for k, v := range s.Refs {
				refs[k] = v
			}
			step["refs"] = refs
		}
		if len(s.Required) > 0 {
			req := make([]interface{}, len(s.Required))
			for j, r := range s.Required {
				req[j] = r
			}
			step["required"] = req
		}
		if len(s.DefaultState) > 0 {
			step["default_state"] = s.DefaultState
		}
		stepsSlice[i] = step
	}

	state := map[string]interface{}{
		"name":        name,
		"description": description,
		"steps":       stepsSlice,
	}
	refs := map[string]interface{}{}
	if authorHash != "" {
		refs["author"] = authorHash
	}
	return Create("observe.template", state, refs)
}

// FromTemplate instantiates a template — creates real blocks from a template pattern.
// values maps step alias to StepOverrides. @alias refs are resolved to previously created block hashes.
func FromTemplate(tmpl TemplateDef, values map[string]StepOverrides) []Block {
	aliases := make(map[string]string)
	var blocks []Block

	for _, step := range tmpl.Steps {
		alias := step.Alias
		if alias == "" {
			alias = step.Type
		}

		overrides := values[alias]

		// Build state from step defaults + overrides
		blockState := make(map[string]interface{})
		for k, v := range step.DefaultState {
			blockState[k] = v
		}
		if overrides.State != nil {
			for k, v := range overrides.State {
				blockState[k] = v
			}
		}

		// Build refs, resolving @aliases
		blockRefs := make(map[string]interface{})
		for role, target := range step.Refs {
			if len(target) > 0 && target[0] == '@' {
				refAlias := target[1:]
				if hash, ok := aliases[refAlias]; ok {
					blockRefs[role] = hash
				}
			} else {
				blockRefs[role] = target
			}
		}
		// Override refs from values
		if overrides.Refs != nil {
			for role, target := range overrides.Refs {
				if len(target) > 0 && target[0] == '@' {
					refAlias := target[1:]
					if hash, ok := aliases[refAlias]; ok {
						blockRefs[role] = hash
					}
				} else {
					blockRefs[role] = target
				}
			}
		}

		block := Create(step.Type, blockState, blockRefs)
		aliases[alias] = block.Hash
		blocks = append(blocks, block)
	}

	return blocks
}
